### PR TITLE
Add command-line support for Duo MFA

### PR DESCRIPTION
**Pull request by kittysyddy**
[WIP] Add command-line support for Duo MFA.

### CODE REVIEW

**Comment by kittysyddy** [comment]
Laundry list:
- Create a wrapper function for Duo API that parses responses (particularly to facilitate graceful exists on errors or if user rejects challenge or timeout)
- Logging statements
- Collect user input in TOTP

@pcmxgti while I do that, are you able to help me shed some light on the file not found error in the test results, I don't understand that

**Comment by pcmxgti** [comment]
> Laundry list:
> 
> * Create a wrapper function for Duo API that parses responses (particularly to facilitate graceful exists on errors or if user rejects challenge or timeout)
> * Logging statements
> * Collect user input in TOTP
> * Some tests
> 
> @pcmxgti while I do that, are you able to help me shed some light on the file not found error in the test results, I don't understand that

Those were fixed in #19. `requirements-dev.txt` was pulling the latest version of `docutils`, and boto has a version boundary `>0.10,<0.16`. When docutils released v0.16 the tests started failing. The interesting part is that boto does not really need docutils anymore, and they have a pending PR for removing that dependency.

**Comment by kittysyddy** [comment]
Hi Ben @hb3b! I have gotten this stable for call, passcode, and push. I don't have access to other factors with Duo but would love your feedback if you use anything else that you would like to see here!

File: tokendito/duo_helpers.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        logging.debug("Non-json response from Duo API: \n{}".format(response))
+
+    if response.status_code >= 400:
+        print("Your Duo authentication has failed with status {}.".format(
+            response.status_code))
+        if json_message and json_message["stat"].lower() != "ok":
+            print("\n{}".format(
+                response.status_code, json_message["message"]))
+        else:
+            print("Please retry with the Debug loglevel to see more information.")
```
Comment (sevignyj) [inline tokendito/duo_helpers.py]: put exact command with parameters to run with debug.

**Comment by hb3b** [comment]
@kittysyddy 👍 Phenomenal job! This works really well. I'd love to see U2F support someday but it's by no means a "breaker" for us. I'll ask more teams here to try your PR and share any issues/feedback. I'm looking forward to seeing this merged to master. It will be a breeze for our employees to setup.

**Comment by wsj-caucus-results** [comment]
> Should add a note that duo is not pre-configurable with this PR yet (will be after env vars PR is done). Also we just make a note that if okta is pre-set then duo will never be offered.

added to secondary readme

File: tokendito/duo_helpers.py (hunk 1 above)
Comment (wsj-caucus-results) [reply tokendito/duo_helpers.py]: done

File: tokendito/okta_helpers.py [hunk 2]
```diff
@@ -115,22 +125,35 @@ def user_mfa_challenge(headers, primary_auth):
...
-    mfa_verify = user_mfa_options(selected_mfa_option,
-                                  headers, mfa_challenge_url, payload, primary_auth)
+
+    if mfa_provider == "duo":
+        payload, headers, callback_url = duo_helpers.authenticate_duo(selected_factor)
+        okta_verify_api_method(callback_url, payload)
+        payload.pop("id", "sig_response")
+        mfa_verify = okta_verify_api_method(
+            mfa_challenge_url, payload, headers)
+    else:
```
[the remark this answers is not in the document]
Comment (wsj-caucus-results) [reply tokendito/okta_helpers.py]: what provider is google authenticator?

**Comment by kittysyddy** [comment]
> Should add a note that duo is not pre-configurable with this PR yet (will be after env vars PR is done). Also we just make a note that if okta is pre-set then duo will never be offered.

added to secondary readme

File: tokendito/duo_helpers.py (hunk 1 above)
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: done

File: tokendito/okta_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kittysyddy) [reply tokendito/okta_helpers.py]: done and added google authenticator

File: docs/README.rst [hunk 3]
```diff
@@ -111,6 +111,13 @@ Additional Usage Reference
                             Note: DEBUG level may display credentials
 
 
+Supported MFA Options:
+""""""""""""""""""""""
+- Native Okta factors (push, call, sms, TOTP)
```
Comment (pcmxgti) [inline docs/README.rst]: Biometrics (FIDO      webauthn) not supported - we should call this out.

File: tokendito/duo_helpers.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    sig_response = duo_factor_callback(duo_info, verify_mfa)
+
+    # Prepare for Okta callback
+    payload = helpers.prepare_payload(id=duo_info["factor_id"],
+                                      sig_response=sig_response,
+                                      stateToken=duo_info["state_token"])
+    headers = {}
+    headers["content-type"] = "application/json"
+    headers["accept"] = "application/json"
+
```
Comment (pcmxgti) [inline tokendito/duo_helpers.py]: Can any of these be verified for failure / accurate response status?

File: tokendito/duo_helpers.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    return payload, headers, duo_info["okta_callback_url"]
```
Comment (pcmxgti) [inline tokendito/duo_helpers.py]: should this be returned as a tuple?

File: tokendito/helpers.py [hunk 6]
```diff
@@ -602,6 +606,103 @@ def update_aws_config(profile, output, region):
...
+    :return passcode:
+    """
+    integer_validation = False
+    try:
+        if int(value):
+            integer_validation = True
+    except ValueError:
+        logging.error("Input is not an integer.")
+
+    return integer_validation
```
Comment (pcmxgti) [inline tokendito/helpers.py]: what happens if the input is a float, or in octal / hex representation? Does it matter? We should probably write up a test for this function.

File: tokendito/helpers.py [hunk 7]
```diff
@@ -602,6 +606,103 @@ def update_aws_config(profile, output, region):
...
+    if entering a TOTP, this does not apply.
+
+    :param valid_range: number of menu options available to user.
+    :return user_input: validated, casted integer from user.
+    """
+    user_input = None
+    while True:
+        user_input = get_input()
+        if validate_input(user_input, valid_range):
+            user_input = int(user_input)
```
Comment (pcmxgti) [inline tokendito/helpers.py]: This should probably use the check_integer function above

File: tokendito/okta_helpers.py [hunk 8]
```diff
@@ -115,22 +125,38 @@ def user_mfa_challenge(headers, primary_auth):
...
-                                  headers, mfa_challenge_url, payload, primary_auth)
+
+    if mfa_provider == "duo":
+        payload, headers, callback_url = duo_helpers.authenticate_duo(
+            selected_factor)
+        okta_verify_api_method(callback_url, payload)
+        payload.pop("id", "sig_response")
+        mfa_verify = okta_verify_api_method(
+            mfa_challenge_url, payload, headers)
+    elif mfa_provider == "okta" or mfa_provider == "google":
```
Comment (pcmxgti) [inline tokendito/okta_helpers.py]: how is FIDO handled here? I'm not sure how yubi keys are represented.
Also, `FIDO:webauthn` should be ignored as an option until supported.

File: tokendito/duo_helpers.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        logging.error(
+            "There was an error connecting to the Duo API: \n{}".format(request_issue))
+
+    json_message = None
+    try:
+        json_message = response.json()
+    except ValueError:
+        logging.debug("Non-json response from Duo API: \n{}".format(response))
+
+    if response.status_code >= 400:
```
[the remark this answers is not in the document]
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: actually there are redirects within the duo workflow that i could not explicitly capture with Burp. i do not know if those would be captured in our requests but i would be more sure that the tool would not break by having the syntax this way

[the remark this answers is not in the document]
Comment (pcmxgti) [reply tokendito/duo_helpers.py]: Shouldn't the success case be as narrow as possible? requests.request() does follow redirects (unless asked not to) before returning data to the caller.

**Comment by sevignyj** [comment]
Screen Shot 2020-01-31 at 1 46 44 PM
 The test gives a false positive (should fail), on top of the error itself.

**Comment by sevignyj** [comment]
I haven't tested end-to-end yet (will add more comments once I do), this looks great overall, could we add more unit tests for the new functionality?

**Comment by sevignyj** [comment]
> Screen Shot 2020-01-31 at 1 46 44 PM
> The test gives a false positive (should fail), on top of the error itself.

replying to my own comment as i realized this was not an error. And this is not an issue with this PR, but more an aesthetic bug that i found while look at test results. We could fix by only logging an error if the profile is not  "expected_failure" (and maybe also rename the profile to "unit_test_expected_failure" for clarify).

File: tokendito/duo_helpers.py (hunk 4 above)
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: all of these functions have response validation inside them. this is the "main" of the duo process so i was trying to make it easy to read the flow: challenge duo mfa, verify that the user completes the challenge, callback/update challenge status to the factor in duo

File: tokendito/duo_helpers.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+    for device in devices:
+        options = soup.find(
+            "fieldset", {"data-device-index": device.split(" - ")[0]})
+        factors = options.findAll("input", {"name": "factor"})
+        for factor in factors:
+            factor_option = {}
+            factor_option["device"] = device
+            factor_option["factor"] = factor["value"]
+            factor_options.append(factor_option)
+    return factor_options
```
[the remark this answers is not in the document]
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: yes there is a catch and exit for if not devices

File: tokendito/duo_helpers.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+
+from bs4 import BeautifulSoup
+from future import standard_library
+import requests
+from tokendito import helpers
+from tokendito import settings
+standard_library.install_aliases()
+
+
+def prepare_duo_info(selected_okta_factor):
```
[the remark this answers is not in the document]
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: the function call is wrapped in a try/except

File: tokendito/duo_helpers.py [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    until the challenge is completed.
+
+    :param duo_info: dict of parameters for Duo.
+    :param mfa_option: the user's selected second factor.
+    :return txid: Duo transaction ID used to track this auth attempt.
+    """
+    url = "https://{}/frame/status".format(duo_info["host"])
+    challenged_mfa = helpers.prepare_payload(txid=txid, sid=duo_info["sid"])
+
+    while True:
```
[the remark this answers is not in the document]
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: done

File: tokendito/duo_helpers.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+        if "status" in verify_mfa:
+            print(verify_mfa["status"])
+
+        try:
+            mfa_result = verify_mfa["result"]
+        except KeyError:
+            logging.debug("No factor result was found"
+                          " in the Duo MFA response: \n{}".format(verify_mfa))
+            continue
```
[the remark this answers is not in the document]
Comment (kittysyddy) [reply tokendito/duo_helpers.py]: this is rewritten

**Pull request closed without merging**